src/main/java/frc/robot/subsystems/shooter/ShooterIO.java

```diff
@@ -8,21 +8,24 @@
 public interface ShooterIO {
     ShooterInputsAutoLogged inputs = new ShooterInputsAutoLogged();
 
-    /**
-     * Sets the position of the hood.
-     *
-     * @param velocity The position of the subsystem.
-     */
-    void setVelocity(MutableMeasure<Velocity<Angle>> velocity);
+    void setTopVelocity(MutableMeasure<Velocity<Angle>> velocity);
+
+    void setBottomVelocity(MutableMeasure<Velocity<Angle>> velocity);
 
     /** Update the inputs of the Shooter */
     void updateInputs();
 
     @AutoLog
     class ShooterInputs {
-        public MutableMeasure<Velocity<Angle>> velocity = MutableMeasure.zero(RotationsPerSecond);
-        public MutableMeasure<Velocity<Angle>> velocitySetpoint =
+        public MutableMeasure<Velocity<Angle>> topVelocity =
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> bottomVelocity =
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> topVelocitySetpoint =
+                MutableMeasure.zero(RotationsPerSecond);
+        public MutableMeasure<Velocity<Angle>> bottomVelocitySetpoint =
                 MutableMeasure.zero(RotationsPerSecond);
-        public MutableMeasure<Voltage> voltage = MutableMeasure.zero(Volts);
+        public MutableMeasure<Voltage> topVoltage = MutableMeasure.zero(Volts);
+        public MutableMeasure<Voltage> bottomVoltage = MutableMeasure.zero(Volts);
```

Comment:
I'm not sure how that would work actually, because the auto log has no way of knowing which set of inputs is which.

Reply:
WDYM?

Reply:
Well it would be easier to have a `bottomRollerInputs` and a `topRollerInputs`, instead of doubling the number of inputs in the class.

Reply:
Ah, I see. I think this might be a nice idea. If we can even have `RollerInputs` class that gets the roller name, our code will be more concise.
We could even create one `Roller` subsystem and instantiate it twice… but that's probably a bit overkill.

Reply:
@rakrakon you can try if you like

Reply:
```
package frc.robot.subsystems.shooter;

import edu.wpi.first.units.Angle;
import edu.wpi.first.units.MutableMeasure;
import edu.wpi.first.units.Velocity;
import edu.wpi.first.units.Voltage;
import org.littletonrobotics.junction.AutoLog;

import static edu.wpi.first.units.Units.RotationsPerSecond;
import static edu.wpi.first.units.Units.Volts;

@AutoLog
public class RollerInputs {
    public String rollerName;
    public MutableMeasure<Velocity<Angle>> velocity =
            MutableMeasure.zero(RotationsPerSecond);
    public MutableMeasure<Velocity<Angle>> velocitySetpoint =
            MutableMeasure.zero(RotationsPerSecond);
    public MutableMeasure<Voltage> voltage = MutableMeasure.zero(Volts);

    public RollerInputs(){

    }

    public RollerInputs(String rollerName){
        this.rollerName = rollerName;
    }

    public String getRollerName(){
        return rollerName;
    }
}

```

```
package frc.robot.subsystems.shooter;

import static edu.wpi.first.units.Units.*;

import edu.wpi.first.units.*;
import org.littletonrobotics.junction.AutoLog;

public interface ShooterIO {
    ShooterInputsAutoLogged inputs = new ShooterInputsAutoLogged();

    void setTopVelocity(MutableMeasure<Velocity<Angle>> velocity);

    void setBottomVelocity(MutableMeasure<Velocity<Angle>> velocity);

    void stop();

    /** Update the inputs of the Shooter */
    void updateInputs();

    @AutoLog
    class ShooterInputs {
        RollerInputs topRoller = new RollerInputs("TopRoller");
        RollerInputs bottomRoller = new RollerInputs("BottomRoller");
    }
}

```

Reply:
1. Note that you should include the language name in a code block, that way Markdown will syntax highlight your code. I edited your comment, you can see what I did in the comment history.
2. I'm gonna need to see the full error, but you might need to duplicate some code and write two inputs classes, one for each roller. You can create an issue about that in the AdvantageKit repository, I guess that's just not a feature they thought of.

Reply:
D:\Users\rakra\robot-2024\build\generated\sources\annotationProcessor\java\main\frc\robot\subsystems\shooter\ShooterInputsAutoLogged.java:11: error: no suitable method found for put(String,RollerInputs)
    table.put("TopRoller", topRoller);
         ^
    method LogTable.put(String,LogValue) is not applicable
      (argument mismatch; RollerInputs cannot be converted to LogValue)
    method LogTable.put(String,byte[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to byte[])
    method LogTable.put(String,boolean) is not applicable
      (argument mismatch; RollerInputs cannot be converted to boolean)
    method LogTable.put(String,int) is not applicable
      (argument mismatch; RollerInputs cannot be converted to int)
    method LogTable.put(String,long) is not applicable
      (argument mismatch; RollerInputs cannot be converted to long)
    method LogTable.put(String,float) is not applicable
      (argument mismatch; RollerInputs cannot be converted to float)
    method LogTable.put(String,double) is not applicable
      (argument mismatch; RollerInputs cannot be converted to double)
    method LogTable.put(String,String) is not applicable
      (argument mismatch; RollerInputs cannot be converted to String)
    method LogTable.<E>put(String,E) is not applicable
      (inference variable E has incompatible bounds
        upper bounds: Enum<E>
        lower bounds: RollerInputs)
    method LogTable.<U>put(String,Measure<U>) is not applicable
      (cannot infer type-variable(s) U
        (argument mismatch; RollerInputs cannot be converted to Measure<U>))
    method LogTable.put(String,boolean[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to boolean[])
    method LogTable.put(String,int[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to int[])
    method LogTable.put(String,long[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to long[])
    method LogTable.put(String,float[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to float[])
    method LogTable.put(String,double[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to double[])
    method LogTable.put(String,String[]) is not applicable
      (argument mismatch; RollerInputs cannot be converted to String[])
    method LogTable.<T#1>put(String,Struct<T#1>,T#1) is not applicable
      (cannot infer type-variable(s) T#1
        (actual and formal argument lists differ in length))
    method LogTable.<T#2>put(String,Struct<T#2>,T#2...) is not applicable
      (cannot infer type-variable(s) T#2
        (argument mismatch; RollerInputs cannot be converted to Struct<T#2>))
    method LogTable.<T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3) is not applicable
      (cannot infer type-variable(s) T#3,MessageType
        (actual and formal argument lists differ in length))
    method LogTable.<T#4>put(String,T#4) is not applicable
      (inference variable T#4 has incompatible bounds
        upper bounds: WPISerializable
        lower bounds: RollerInputs)
    method LogTable.<T#5>put(String,T#5...) is not applicable
      (inference variable T#5 has incompatible bounds
        upper bounds: StructSerializable
        lower bounds: RollerInputs)
  where E,U,T#1,T#2,T#3,MessageType,T#4,T#5 are type-variables:
    E extends Enum<E> declared in method <E>put(String,E)
    U extends Unit<U> declared in method <U>put(String,Measure<U>)
    T#1 extends Object declared in method <T#1>put(String,Struct<T#1>,T#1)
    T#2 extends Object declared in method <T#2>put(String,Struct<T#2>,T#2...)
    T#3 extends Object declared in method <T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3)
    MessageType extends ProtoMessage<?> declared in method <T#3,MessageType>put(String,Protobuf<T#3,MessageType>,T#3)
    T#4 extends WPISerializable declared in method <T#4>put(String,T#4)
    T#5 extends StructSerializable declared in method <T#5>put(String,T#5...)

Reply:
Yep:
```
no suitable method found for put(String,RollerInputs)
```
Not supported for classes. But making two inputs classes should work.

Reply:
fe871ab
0393153
d7058f6

Reply:
That's a better solution than I thought of! Did you check and see if you really get two different inputs sets?